### ISSUE TITLE
[Performance] Track direct Triton JIT host-launch latency and add a regression benchmark

### ISSUE THREAD

**Issue by rogerchen838** [triton-lang/triton-ascend#1675]
## Summary

The legacy steady-state Triton C launcher on `release/3.2.2-dev` has a large
host enqueue cost for a tiny direct `@triton.jit` kernel.  On Ascend 910B2,
the legacy launcher takes **23.654 us** per enqueue versus **12.142 us** for
eager `torch.add` (1.95x slower).

An optimized `release/3.2.2-dev` launch implementation reduces the same
direct C-launcher benchmark to **8.599 us** (2.75x lower than the legacy
launcher and 1.35x faster than eager).  This makes host launch time a material
optimization opportunity and warrants a small direct-JIT regression benchmark.

This report is related to the implementation work in #1651.  It provides a
focused device-side acceptance benchmark; it does not prescribe the exact
implementation, and does not claim that the unmeasured GitHub `main` launcher
has the same legacy number.

## What is measured

This is **host enqueue time**, not NPU execution time:

1. Compile a single direct `@triton.jit` elementwise-add kernel once.
2. Obtain the resulting `CompiledKernel`'s already-built C launcher.
3. Precompute its stream and launch metadata once.
4. In the timed loop call the C launcher directly for 5,000 launches.
5. Stop the timer before the final `torch.npu.synchronize()`.

The benchmark deliberately does not call `JITFunction.run()` in the timed
loop, because Python-side argument binding and compile-cache lookup are not
kernel-launch overhead.  It also does not use `torch.compile` or Inductor.

The test uses a direct add of 400 FP32 elements, grid `(1,)`, `BLOCK=512`.
Before timing, the compiled launcher output is checked against `x + y`.
Each reported value is the median of five independent batches; every batch has
1,000 untimed warmup launches followed by 5,000 timed enqueues and a
synchronization after the timer stops.

## Reproducer core

```python
import statistics
import time

import torch
import torch_npu  # noqa: F401
import triton
import triton.language as tl


@triton.jit
def launch_probe(x_ptr, y_ptr, out_ptr, numel, BLOCK: tl.constexpr):
    offsets = tl.arange(0, BLOCK)
    mask = offsets < numel
    x = tl.load(x_ptr + offsets, mask=mask)
    y = tl.load(y_ptr + offsets, mask=mask)
    tl.store(out_ptr + offsets, x + y, mask=mask)


def compiled_c_launcher(x, y, output, numel):
    # Compile only once.  This occurs before the timed section.
    compiled = launch_probe.run(
        x, y, output, numel, BLOCK=512, grid=(1,), warmup=False
    )
    compiled._init_handles()

    driver = triton.runtime.driver.active
    stream = driver.get_current_stream(driver.get_current_device())
    metadata = compiled.launch_metadata((1,), stream, x, y, output, numel)
    fixed_args = (
        1, 1, 1, stream, compiled.function, compiled.packed_metadata, metadata,
        compiled.__class__.launch_enter_hook,
        compiled.__class__.launch_exit_hook,
    )
    kernel_args = (x, y, output, numel)

    # `launch` is the compiled extension's raw C entry, not JITFunction.run.
    return lambda: compiled.run.launch(*fixed_args, *kernel_args)


def measure_enqueue(call, warmup=1000, iters=5000, repeats=5):
    samples_us = []
    for _ in range(repeats):
        for _ in range(warmup):
            call()
        torch.npu.synchronize()
        start = time.perf_counter()
        for _ in range(iters):
            call()
        elapsed = time.perf_counter() - start  # final sync intentionally excluded
        torch.npu.synchronize()
        samples_us.append(elapsed * 1e6 / iters)
    return statistics.median(samples_us), samples_us


x = torch.randn((4, 100), device="npu", dtype=torch.float32)
y = torch.randn((4, 100), device="npu", dtype=torch.float32)
out = torch.empty_like(x)
triton_call = compiled_c_launcher(x, y, out, 400)
triton_call(); torch.npu.synchronize()
torch.testing.assert_close(out, x + y, rtol=1e-3, atol=1e-3)

triton_us, triton_samples = measure_enqueue(triton_call)
eager_us, eager_samples = measure_enqueue(lambda: torch.add(x, y, out=out))
print(triton_us, triton_samples)
print(eager_us, eager_samples)
```

The full script also supports launching the current optimized implementation
in a separate process and selecting `auto` (FASTCALL when available) or the
legacy raw C entry.  Separate processes are required because Triton's active
backend driver is process-global.

## Results

All values are median host enqueue time in microseconds per launch.

| implementation / launcher entry | Triton | eager `torch.add` | Triton vs eager |
|---|---:|---:|---:|
| legacy `release/3.2.2-dev` C launcher (`METH_VARARGS`) | 23.654 | 12.142 | 1.948x slower |
| optimized `release/3.2.2-dev`, raw `launch` (`METH_VARARGS`) | 8.633 | 10.669 | 1.236x faster |
| optimized `release/3.2.2-dev`, `fast_launch` (`METH_FASTCALL`) | 8.599 | 11.655 | 1.355x faster |

Raw per-batch samples:

```text
legacy Triton:     23.654, 24.138, 23.529, 23.476, 23.750 us
legacy eager:      14.474, 12.141, 12.093, 12.142, 12.219 us

optimized FASTCALL Triton:
                    9.675, 8.500, 8.566, 8.633, 8.599 us
optimized eager:   11.673, 11.654, 11.592, 11.655, 11.738 us
```

The direct C launch improved from 23.654 us to 8.599 us (2.75x).  In this
particular microbenchmark, forcing the optimized raw `METH_VARARGS` entry is
8.633 us, so FASTCALL itself accounts for only about 0.4%; the dominant win is
the optimized ordered launch/capture path, not Python argument parsing alone.

## Request

1. Retain or port a low-overhead, ordering-preserving launch path for direct
   Triton JIT launches, and evaluate #1651 against this benchmark.
2. Add a direct-JIT host-launch microbenchmark (or an equivalent regression
   test) that compiles outside timing and calls the compiled C launcher in the
   timed loop.  It should report host enqueue latency separately from NPU
   kernel duration.
3. Document which entry points and instrumentation modes are covered by the
   low-overhead path.  The fast path should retain the fallback behavior for
   launch hooks/profiling where necessary.

## Environment

- Hardware: 8 x Huawei Ascend 910B2; benchmark ran on NPU 0 only
  - Product: `IT21HMDC_Bin0`; 24 AICores; 65,536 MiB HBM per card
- Driver / board software (`npu-smi`): `26.0.rc1`
- Firmware: `9.0.0.0.205`
- CANN toolkit and operator packages: `9.1.0-beta.3`
- Architecture: AArch64
- Container OS: Ubuntu 22.04.5 LTS
- Host kernel visible in the container:
  `6.6.0-132.0.0.111.oe2403sp3.aarch64`
- glibc: 2.35
- Python: 3.10.20 (conda environment)
- PyTorch: `2.13.0+cpu`
- torch-npu: `2.13.0+git4d712d1`
- Triton-Ascend branch: `release/3.2.2-dev`

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.
